src/porepy/fracs/wells_3d.py

```diff
@@ -436,6 +440,110 @@ def compute_well_fracture_intersections(
+    gs_w = gb.grids_of_dimension(dim_max - 2)
+
+    # Pre-compute some well informations
+    nodes_w = np.empty(gs_w.size, dtype=object)
+    for idw, g_w in enumerate(gs_w):
+        g_w_cn = g_w.cell_nodes()
+        g_w_cells = np.arange(g_w.num_cells)
+        # get the cells of the 0d as segments (start, end)
+        first = g_w_cn.indptr[g_w_cells]
+        second = g_w_cn.indptr[g_w_cells + 1]
+
+        nodes_w[idw] = (
+            g_w_cn.indices[pp.utils.mcolon.mcolon(first, second)].reshape((-1, 2)).T
+        )
+
+    # Operate on the rock matrix grid
+    (faces, cells, _) = sps.find(g_max.cell_faces)
+    faces = faces[np.argsort(cells)]
+
+    nodes, _, _ = sps.find(g_max.face_nodes)
+    indptr = g_max.face_nodes.indptr
+
+    # Loop on all the well grids
+    for g_w, n_w in zip(gs_w, nodes_w):
+        # extract the start and end point of the segments
+        start = g_w.nodes[:, n_w[0]]
+        end = g_w.nodes[:, n_w[1]]
+
+        # Lists for the cell_cell_map
+        primary_to_mortar_I, primary_to_mortar_J, primary_to_mortar_data = [], [], []
+
+        # Operate on the segments
+        for seg_id, (seg_start, seg_end) in enumerate(zip(start.T, end.T)):
+            # Create the box for the segment by ordering its start and end
+            box = np.sort(np.vstack((seg_start, seg_end)), axis=0).flatten()
+            seg_cells = tree.search(pp.adtree.ADTNode("dummy_node", box))
+
+            # Loop on all the higher dimensional cells
+            for c in seg_cells:
+                # For the current cell retrieve its faces
+                loc = slice(g_max.cell_faces.indptr[c], g_max.cell_faces.indptr[c + 1])
+                faces_loc = faces[loc]
+                # Get the local nodes, face based
+                poly = np.array(
+                    [
+                        g_max.nodes[:, nodes[indptr[f] : indptr[f + 1]]]
+                        for f in faces_loc
+                    ]
+                )
+                # Compute the intersections between the segment and the current higher
+                # dimensional cell
+                ratio = pp.intersections.segments_polyhedron(
+                    seg_start, seg_end, poly, tol
+                )
+                # Store the requested information to build the projection operator
+                if ratio > 0:
+                    primary_to_mortar_I += [seg_id]
+                    primary_to_mortar_J += [c]
+                    primary_to_mortar_data += ratio.tolist()
+        primary_to_mortar_int = sps.csc_matrix(
+            (primary_to_mortar_data, (primary_to_mortar_I, primary_to_mortar_J)),
+            shape=(g_w.num_cells, g_max.num_cells),
+        )
+        secondary_to_mortar_int = sps.diags(np.ones(g_w.num_cells), format="csc")
+
+        # create the mortar grid and set the maps
+        side_g = {pp.grids.mortar_grid.MortarSides.LEFT_SIDE: g_w.copy()}
+        mg = pp.MortarGrid(g_w.dim, side_g, codim=g_max.dim - g_w.dim)
```

Comment:
okay, I've changed removing the new function in the MortarGrid

Reply:
if I give the primary_secondary_map then the constructor calls the method _init_projections which creates problems since it assumes a matching coupling between the primary and secondary grids. in this way I don't have this problem